File: Archiv/debug_trace_intermediate_values.m
% Debug script to trace intermediate values in linReach_adaptive
% This tracks all intermediate values step by step to compare with Python

addpath(genpath('cora_matlab'));

% Setup test case - use a simpler case that we can trace
params.tStart = 0;
params.tFinal = 10;
params.R0 = zonotope([0; 0], 0.1*eye(2));
params.U = zonotope(0, 0);

options = struct();
options.alg = 'lin';
options.tensorOrder = 3;
options.timeStep = 0.1;
options.taylorTerms = 10;
options.zonotopeOrder = 50;
options.reductionTechnique = 'adaptive';
options.redFactor = 0.9;
options.decrFactor = 0.5;
options.minorder = 1;
options.maxError = 0.1 * ones(2, 1);
options.zetaphi = [0.5; 0.5];
options.zetaK = 0.1;
options.orders = ones(2, 1);
options.i = 1;
options.error_adm_horizon = 1e-10 * ones(2, 1);  % Start with small value

% Create simple 2D system
f = @(x, u) [-x(1) + x(2); -x(2) + u];
sys = nonlinearSys('simple', f, 2, 1);

fprintf('=== Tracing Intermediate Values in linReach_adaptive ===\n\n');

% Open file for detailed logging
fid = fopen('matlab_intermediate_values.txt', 'w');
fprintf(fid, '=== MATLAB Intermediate Values Trace ===\n\n');

try
    % Simulate one iteration of the inner loop
    error_adm = options.error_adm_horizon;
    fprintf('Step 0: Initial error_adm_horizon\n');
    fprintf(fid, 'Step 0: Initial error_adm_horizon\n');
    fprintf('  error_adm = [%.15e; %.15e]\n', error_adm(1), error_adm(2));
    fprintf(fid, '  error_adm = [%.15e; %.15e]\n', error_adm(1), error_adm(2));
    
    % Create Verror
    fprintf('\nStep 1: Create Verror from error_adm\n');
    fprintf(fid, '\nStep 1: Create Verror from error_adm\n');
    errG = diag(error_adm);
    Verror = zonotope(0*error_adm, errG(:, any(errG, 1)));
    Verror_center = center(Verror);
    Verror_gens = generators(Verror);
    fprintf('  Verror center: [%.15e; %.15e]\n', Verror_center(1), Verror_center(2));
    fprintf(fid, '  Verror center: [%.15e; %.15e]\n', Verror_center(1), Verror_center(2));
    fprintf('  Verror generators shape: %dx%d\n', size(Verror_gens, 1), size(Verror_gens, 2));
    fprintf(fid, '  Verror generators shape: %dx%d\n', size(Verror_gens, 1), size(Verror_gens, 2));
    fprintf('  Verror generators max abs: %.15e\n', max(max(abs(Verror_gens))));
    fprintf(fid, '  Verror generators max abs: %.15e\n', max(max(abs(Verror_gens))));
    
    % For a real trace, we'd need to call the actual functions
    % But for now, let's document what values we expect at each step
    
    fprintf('\n=== Expected Computation Flow ===\n');
    fprintf(fid, '\n=== Expected Computation Flow ===\n');
    fprintf('1. error_adm -> Verror (zonotope with diag(error_adm))\n');
    fprintf('2. Verror -> RallError (via errorSolution_adaptive)\n');
    fprintf('3. Rlinti + RallError -> Rmax\n');
    fprintf('4. Rmax -> Rred (reduced)\n');
    fprintf('5. Rred, U -> Z (cartProd)\n');
    fprintf('6. Z, H -> errorSec (0.5 * quadMap)\n');
    fprintf('7. errorSec -> VerrorDyn (after reduction)\n');
    fprintf('8. VerrorDyn -> trueError (abs(center) + sum(abs(generators)))\n');
    fprintf('9. trueError, error_adm -> perfIndCurr (max(trueError ./ error_adm))\n');
    fprintf('10. perfIndCurr -> convergence check\n');
    
    fprintf(fid, '1. error_adm -> Verror (zonotope with diag(error_adm))\n');
    fprintf(fid, '2. Verror -> RallError (via errorSolution_adaptive)\n');
    fprintf(fid, '3. Rlinti + RallError -> Rmax\n');
    fprintf(fid, '4. Rmax -> Rred (reduced)\n');
    fprintf(fid, '5. Rred, U -> Z (cartProd)\n');
    fprintf(fid, '6. Z, H -> errorSec (0.5 * quadMap)\n');
    fprintf(fid, '7. errorSec -> VerrorDyn (after reduction)\n');
    fprintf(fid, '8. VerrorDyn -> trueError (abs(center) + sum(abs(generators)))\n');
    fprintf(fid, '9. trueError, error_adm -> perfIndCurr (max(trueError ./ error_adm))\n');
    fprintf(fid, '10. perfIndCurr -> convergence check\n');
    
    fprintf('\n=== Key Values to Track ===\n');
    fprintf(fid, '\n=== Key Values to Track ===\n');
    fprintf('At each inner loop iteration:\n');
    fprintf('  - error_adm (input)\n');
    fprintf('  - RallError radius (max sum of abs generators)\n');
    fprintf('  - Rmax radius (max sum of abs generators)\n');
    fprintf('  - Z radius (max sum of abs generators)\n');
    fprintf('  - errorSec radius (max sum of abs generators)\n');
    fprintf('  - VerrorDyn radius (max sum of abs generators)\n');
    fprintf('  - trueError (vector)\n');
    fprintf('  - perfIndCurr (scalar)\n');
    fprintf('  - perfInds (array of perfIndCurr values)\n');
    
    fprintf(fid, 'At each inner loop iteration:\n');
    fprintf(fid, '  - error_adm (input)\n');
    fprintf(fid, '  - RallError radius (max sum of abs generators)\n');
    fprintf(fid, '  - Rmax radius (max sum of abs generators)\n');
    fprintf(fid, '  - Z radius (max sum of abs generators)\n');
    fprintf(fid, '  - errorSec radius (max sum of abs generators)\n');
    fprintf(fid, '  - VerrorDyn radius (max sum of abs generators)\n');
    fprintf(fid, '  - trueError (vector)\n');
    fprintf(fid, '  - perfIndCurr (scalar)\n');
    fprintf(fid, '  - perfInds (array of perfIndCurr values)\n');
    
catch ME
    fprintf('\n=== ERROR ===\n');
    fprintf('Identifier: %s\n', ME.identifier);
    fprintf('Message: %s\n', ME.message);
    fprintf(fid, '\n=== ERROR ===\n');
    fprintf(fid, 'Identifier: %s\n', ME.identifier);
    fprintf(fid, 'Message: %s\n', ME.message);
end

fclose(fid);
fprintf('\n=== Trace complete. Values saved to matlab_intermediate_values.txt ===\n');

File: Archiv/debug_trace_intermediate_values.py
"""
Debug script to trace intermediate values in linReach_adaptive (Python)
This tracks all intermediate values step by step to compare with MATLAB
"""

import sys
import os
sys.path.insert(0, os.path.abspath('.'))

import numpy as np
from cora_python.contSet.zonotope import Zonotope
from cora_python.contDynamics.nonlinearSys.nonlinearSys import NonlinearSys

print('=== Tracing Intermediate Values in linReach_adaptive (Python) ===\n')

# Open file for detailed logging
with open('python_intermediate_values.txt', 'w') as f:
    f.write('=== Python Intermediate Values Trace ===\n\n')
    
    # Setup test case - use a simpler case that we can trace
    params = {
        'tStart': 0,
        'tFinal': 10,
        'R0': Zonotope(np.zeros((2, 1)), 0.1 * np.eye(2)),
        'U': Zonotope(np.zeros((1, 1)), np.zeros((1, 1)))
    }
    
    options = {
        'alg': 'lin',
        'tensorOrder': 3,
        'timeStep': 0.1,
        'taylorTerms': 10,
        'zonotopeOrder': 50,
        'reductionTechnique': 'adaptive',
        'redFactor': 0.9,
        'decrFactor': 0.5,
        'minorder': 1,
        'maxError': 0.1 * np.ones((2, 1)),
        'zetaphi': [0.5, 0.5],
        'zetaK': 0.1,
        'orders': np.ones((2, 1)),
        'i': 1,
        'error_adm_horizon': 1e-10 * np.ones((2, 1))  # Start with small value
    }
    
    try:
        # Simulate one iteration of the inner loop
        error_adm = options['error_adm_horizon']
        print('Step 0: Initial error_adm_horizon')
        f.write('Step 0: Initial error_adm_horizon\n')
        print(f"  error_adm = [{error_adm[0, 0]:.15e}; {error_adm[1, 0]:.15e}]")
        f.write(f"  error_adm = [{error_adm[0, 0]:.15e}; {error_adm[1, 0]:.15e}]\n")
        
        # Create Verror
        print('\nStep 1: Create Verror from error_adm')
        f.write('\nStep 1: Create Verror from error_adm\n')
        errG = np.diag(error_adm.flatten())
        Verror = Zonotope(np.zeros_like(error_adm), errG[:, np.any(errG, axis=0)])
        Verror_center = Verror.center()
        Verror_gens = Verror.generators()
        print(f"  Verror center: [{Verror_center[0, 0]:.15e}; {Verror_center[1, 0]:.15e}]")
        f.write(f"  Verror center: [{Verror_center[0, 0]:.15e}; {Verror_center[1, 0]:.15e}]\n")
        print(f"  Verror generators shape: {Verror_gens.shape}")
        f.write(f"  Verror generators shape: {Verror_gens.shape}\n")
        print(f"  Verror generators max abs: {np.max(np.abs(Verror_gens)):.15e}")
        f.write(f"  Verror generators max abs: {np.max(np.abs(Verror_gens)):.15e}\n")
        
        print('\n=== Expected Computation Flow ===')
        f.write('\n=== Expected Computation Flow ===\n')
        print('1. error_adm -> Verror (zonotope with diag(error_adm))')
        print('2. Verror -> RallError (via errorSolution_adaptive)')
        print('3. Rlinti + RallError -> Rmax')
        print('4. Rmax -> Rred (reduced)')
        print('5. Rred, U -> Z (cartProd)')
        print('6. Z, H -> errorSec (0.5 * quadMap)')
        print('7. errorSec -> VerrorDyn (after reduction)')
        print('8. VerrorDyn -> trueError (abs(center) + sum(abs(generators)))')
        print('9. trueError, error_adm -> perfIndCurr (max(trueError ./ error_adm))')
        print('10. perfIndCurr -> convergence check')
        
        f.write('1. error_adm -> Verror (zonotope with diag(error_adm))\n')
        f.write('2. Verror -> RallError (via errorSolution_adaptive)\n')
        f.write('3. Rlinti + RallError -> Rmax\n')
        f.write('4. Rmax -> Rred (reduced)\n')
        f.write('5. Rred, U -> Z (cartProd)\n')
        f.write('6. Z, H -> errorSec (0.5 * quadMap)\n')
        f.write('7. errorSec -> VerrorDyn (after reduction)\n')
        f.write('8. VerrorDyn -> trueError (abs(center) + sum(abs(generators)))\n')
        f.write('9. trueError, error_adm -> perfIndCurr (max(trueError ./ error_adm))\n')
        f.write('10. perfIndCurr -> convergence check\n')
        
        print('\n=== Key Values to Track ===')
        f.write('\n=== Key Values to Track ===\n')
        print('At each inner loop iteration:')
        print('  - error_adm (input)')
        print('  - RallError radius (max sum of abs generators)')
        print('  - Rmax radius (max sum of abs generators)')
        print('  - Z radius (max sum of abs generators)')
        print('  - errorSec radius (max sum of abs generators)')
        print('  - VerrorDyn radius (max sum of abs generators)')
        print('  - trueError (vector)')
        print('  - perfIndCurr (scalar)')
        print('  - perfInds (array of perfIndCurr values)')
        
        f.write('At each inner loop iteration:\n')
        f.write('  - error_adm (input)\n')
        f.write('  - RallError radius (max sum of abs generators)\n')
        f.write('  - Rmax radius (max sum of abs generators)\n')
        f.write('  - Z radius (max sum of abs generators)\n')
        f.write('  - errorSec radius (max sum of abs generators)\n')
        f.write('  - VerrorDyn radius (max sum of abs generators)\n')
        f.write('  - trueError (vector)\n')
        f.write('  - perfIndCurr (scalar)\n')
        f.write('  - perfInds (array of perfIndCurr values)\n')
        
    except Exception as e:
        print(f'\n=== ERROR ===')
        print(f'Type: {type(e).__name__}')
        print(f'Message: {str(e)}')
        f.write(f'\n=== ERROR ===\n')
        f.write(f'Type: {type(e).__name__}\n')
        f.write(f'Message: {str(e)}\n')
        import traceback
        traceback.print_exc()
        f.write(traceback.format_exc())

print('\n=== Trace complete. Values saved to python_intermediate_values.txt ===')
